File: backend/app/engine/depth.py
"""Задание на глубину (раздел 6, доп. задача): вертикальный профиль трассы.

Режим включается ENGINE_DEPTH_MODE=1. В основной двумерной задаче (§8.1)
глубины не вычисляются — depth_start/depth_end = null, Kгл = 1.

Правила раздела 6:
  - базовая глубина до верха расчётного габарита — 3,0 м, минимум 0,7 м;
  - отметки дискретизируются шагом 0,5 м;
  - продольный уклон смены глубины ≤ 0,10 м/м (пандусы между отметками);
  - на точечных пересечениях — горизонтальная площадка 4,0 м;
  - газ/кабель/теплосеть: проход ВЫШЕ или НИЖЕ — выбирается более дешёвый
    (с учётом Kгл), с минимальным вертикальным просветом из таблицы 5.1;
  - Kгл = 1 + 0,10·(h − 3) при h > 3 м (при h ≤ 3 м Kгл = 1);
  - в точках пересечения профилем отметки 3,0 м ставится технический узел,
    участок делится на части.
"""

from __future__ import annotations

import math

from shapely.geometry import LineString, Point

from .hydraulics import NewSegment, TechnicalNode, subline
from .refdata import RefData


def quantize(depth_m: float, ref: RefData) -> float:
    """Квантование отметки шагом 0,5 м с соблюдением минимума 0,7 м."""
    step = ref.depth_rule("step_m", 0.5)
    dmin = ref.depth_rule("min_depth_m", 0.7)
    return max(dmin, round(round(depth_m / step) * step, 3))


def target_depth_m(ref: RefData) -> float:
    """Целевая (базовая) глубина заложения, квантованная шагом."""
    return quantize(ref.depth_rule("base_depth_m", 3.0), ref)


def chamber_depth_m(ref: RefData, connecting_dn_mm: float | None = None) -> float:
    """Отметка дна камеры: базовая глубина + высота габарита + запас."""
    h = ref.envelope_height_m(connecting_dn_mm) if connecting_dn_mm else 0.0
    return quantize(ref.depth_rule("base_depth_m", 3.0) + h + 0.25, ref)


def _crossing_constraints(seg: NewSegment, ref: RefData,
                          special_zones: list) -> list[tuple[float, float, float]]:
    """Требуемые отметки на пересечениях: [(start_m, end_m, depth_m)].

    Для газопровода/кабеля/теплосети — выбор «выше или ниже» по цене
    (дешевле тот вариант, у которого Kгл меньше). Для дороги/трамвая
    (под объектом) базовая глубина 3,0 м уже удовлетворяет просвету —
    ограничение добавляется только если просвет не выполняется.
    """
    line = LineString(seg.coords)
    length = line.length
    base = target_depth_m(ref)
    flat = ref.depth_rule("crossing_flat_m", 4.0)
    out: list[tuple[float, float, float]] = []
    env_h = ref.envelope_height_m(seg.diameter_mm)

    for z in special_zones:
        if not line.intersects(z.geom):
            continue
        rule = ref.restriction_rule(z.restriction_type) or {}
        vert = rule.get("vertical_clearance_m")
        under = rule.get("under_clearance_m")
        if vert is None and under is None:
            continue
        inter = line.intersection(z.geom)
        ds = [line.project(Point(p)) for p in _points_of(inter)]
        if not ds:
            continue
        d = (min(ds) + max(ds)) / 2.0
        if under is not None:
            need = quantize(float(under) + 0.0, ref)  # верх габарита глубже просвета
            if need > base:
                out.append((max(0.0, d - flat / 2), min(length, d + flat / 2), need))
            continue
        # выше или ниже существующей коммуникации — что дешевле
        exist_d = float(ref.depth.get("existing_depth_m", {}).get(
            z.restriction_type or "", 1.5))
        above = quantize(exist_d - float(vert) - env_h, ref)
        below = quantize(exist_d + float(vert) + env_h, ref)
        chosen = above if ref.depth_k(above) <= ref.depth_k(below) else below
        if abs(chosen - base) > 1e-9:
            out.append((max(0.0, d - flat / 2), min(length, d + flat / 2), chosen))
    return out


def _points_of(geom) -> list[tuple[float, float]]:
    t = geom.geom_type
    if t == "Point":
        return [(geom.x, geom.y)]
    if t == "MultiPoint":
        return [(p.x, p.y) for p in geom.geoms]
    if t == "LineString":
        return list(geom.coords)
    if t in ("MultiLineString", "GeometryCollection"):
        pts: list[tuple[float, float]] = []
        for g in geom.geoms:
            pts.extend(_points_of(g))
        return pts
    return [(geom.representative_point().x, geom.representative_point().y)]


def _profile(length: float, constraints: list[tuple[float, float, float]],
             entry_depth_m: float | None, ref: RefData) -> list[tuple[float, float]]:
    """Продольный профиль: [(distance_m, depth_m)] с пандусами ≤ max_slope.

    Требуемые отметки соединяются с базовой глубиной пандусами
    ограниченного уклона; результат квантуется шагом 0,5 м.
    """
    base = target_depth_m(ref)
    slope = max(ref.depth_rule("max_slope", 0.10), 1e-6)
    step = 2.0
    n = max(2, int(length / step) + 1)
    xs = [i * length / (n - 1) for i in range(n)]
    # целевая отметка: база, жёсткие ограничения пересечений, входная отметка
    target = [base] * n
    if entry_depth_m is not None:
        target[0] = quantize(entry_depth_m, ref)
    for a, b, h in constraints:
        for i, x in enumerate(xs):
            if a - 1e-9 <= x <= b + 1e-9:
                if h > base:
                    target[i] = max(target[i], h)
                else:
                    target[i] = h if target[i] == base else min(target[i], h)

    # сглаживание уклоном: прямой и обратный проход
    prof = list(target)
    for _ in range(4):
        for i in range(1, n):
            prof[i] = min(prof[i], prof[i - 1] + slope * (xs[i] - xs[i - 1])) \
                if prof[i] > prof[i - 1] else max(prof[i], prof[i - 1] - slope * (xs[i] - xs[i - 1]))
        for i in range(n - 2, -1, -1):
            prof[i] = min(prof[i], prof[i + 1] + slope * (xs[i + 1] - xs[i])) \
                if prof[i] > prof[i + 1] else max(prof[i], prof[i + 1] - slope * (xs[i + 1] - xs[i]))
        # ограничения — жёсткие: восстанавливаем после сглаживания
        for a, b, h in constraints:
            for i, x in enumerate(xs):
                if a - 1e-9 <= x <= b + 1e-9:
                    if h > base:
                        prof[i] = max(prof[i], h)
                    else:
                        prof[i] = min(prof[i], h)
    return [(x, quantize(h, ref)) for x, h in zip(xs, prof)]


def apply_depth(seg: NewSegment, ref: RefData, seq: dict,
                tech_nodes: list[TechnicalNode], special_zones: list,
                entry_depth_m: float | None = None) -> list[NewSegment]:
    """Построить профиль §6, разрезать участок на отметке 3,0 м, посчитать Kгл.

    Возвращает список подучастков с depth_start/depth_end и coords3d;
    в точках пересечения профилем отметки free_depth ставятся техузлы.
    """
    line = LineString(seg.coords)
    length = line.length
    base = target_depth_m(ref)
    constraints = _crossing_constraints(seg, ref, special_zones)
    prof = _profile(length, constraints, entry_depth_m, ref)

    # точки деления: где профиль пересекает базовую отметку (смена режима)
    cuts = [0.0]
    for i in range(1, len(prof)):
        h0, h1 = prof[i - 1][1], prof[i][1]
        if (h0 - base) * (h1 - base) < 0:
            x0, x1 = prof[i - 1][0], prof[i][0]
            t = (base - h0) / (h1 - h0)
            cuts.append(x0 + t * (x1 - x0))
    cuts.append(length)
    cuts = sorted(set(round(c, 2) for c in cuts))

    parts: list[NewSegment] = []
    cur_start_node = seg.start_node_id
    for k in range(len(cuts) - 1):
        a, b = cuts[k], cuts[k + 1]
        if b - a < 0.01:
            continue
        piece = subline(line, a, b)
        # глубины на концах куска — интерполяция профиля
        d_a = _depth_at(prof, a)
        d_b = _depth_at(prof, b)
        is_last = k == len(cuts) - 2
        if is_last:
            end_node = seg.end_node_id
        else:
            seq["tech"] += 1
            end_node = f"node-{seq['tech']}"
            tech_nodes.append(TechnicalNode(
                object_id=end_node,
                point=line.interpolate(b),
                reason="пересечение отметки 3,0 м (раздел 6)"))
        part = NewSegment(
            object_id=seg.object_id if k == 0 and is_last else f"{seg.object_id}.{k + 1}",
            coords=list(piece.coords),
            flow_tph=seg.flow_tph,
            role=seg.role,
            method=seg.method,
            k_special=seg.k_special,
            start_node_id=cur_start_node,
            end_node_id=end_node,
            depth_start=round(d_a, 2),
            depth_end=round(d_b, 2),
        )
        part.diameter_mm = seg.diameter_mm
        part.warnings = list(seg.warnings)
        k_avg = (ref.depth_k(d_a) + ref.depth_k(d_b)) / 2.0
        part._tariff_rub_m = seg._tariff_rub_m
        part.cost_rub = seg.cost_rub * (part.length_m / max(seg.length_m, 1e-9)) * k_avg
        part.coords3d = _coords3d(piece, prof, a, b)
        parts.append(part)
        cur_start_node = end_node

    if not parts:  # вырожденный случай — участок как есть
        seg.depth_start = seg.depth_end = base
        seg.coords3d = [(x, y, -base) for x, y in seg.coords]
        parts = [seg]
    return parts


def _depth_at(prof: list[tuple[float, float]], x: float) -> float:
    """Глубина профиля в точке x (линейная интерполяция)."""
    if x <= prof[0][0]:
        return prof[0][1]
    if x >= prof[-1][0]:
        return prof[-1][1]
    for i in range(1, len(prof)):
        if prof[i][0] >= x:
            x0, h0 = prof[i - 1]
            x1, h1 = prof[i]
            t = (x - x0) / max(x1 - x0, 1e-9)
            return h0 + t * (h1 - h0)
    return prof[-1][1]


def _coords3d(piece: LineString, prof, a: float, b: float) -> list:
    """3D-координаты куска: Z = −глубина (м, ниже поверхности)."""
    out = []
    L = piece.length
    for x, y in piece.coords:
        d = piece.project(Point(x, y))
        out.append((x, y, -_depth_at(prof, a + (d if L > 1e-9 else 0.0))))
    return out

File: app/src/main/java/ru/heatnet/dit/engine/core/Depth.java
package ru.heatnet.dit.engine.core;

import org.locationtech.jts.geom.Coordinate;
import org.locationtech.jts.geom.Geometry;
import org.locationtech.jts.geom.GeometryFactory;
import org.locationtech.jts.geom.LineString;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeSet;

/**
 * Дополнительный режим с учётом глубины (§5 техприложения).
 *
 * Глубина — до ВЕРХА расчётного габарита. Обычная 3,0 м, минимальная 0,7 м,
 * максимальная не ограничена, дискретный шаг не задаётся (разъяснение №16).
 * Уклон спуска/подъёма ≤ 0,10 м/м. Kгл = 1 при h ≤ 3,0 м, иначе
 * Kгл = 1 + 0,10·(h − 3); на равномерном спуске/подъёме — среднее
 * арифметическое Kгл на концах. Пересечение отметки 3,0 м — технический
 * узел и деление участка. Z-координаты не формируются — только атрибуты
 * depth_start/depth_end.
 *
 * Режим включается ENGINE_DEPTH_MODE=1. В базовом 2D-режиме глубины null,
 * Kгл = 1.
 */
public final class Depth {

    private static final GeometryFactory GF = new GeometryFactory();

    private Depth() {
    }

    /** Приведение отметки: не ниже минимума 0,7 м, округление до сантиметра. */
    public static double quantize(double depthM, RefData ref) {
        double dmin = ref.depthRule("min_depth_m", 0.7);
        return Math.max(dmin, Math.rint(depthM * 100.0) / 100.0);
    }

    /** Базовая глубина заложения (до верха габарита). */
    public static double targetDepthM(RefData ref) {
        return ref.depthRule("base_depth_m", 3.0);
    }

    // типы ограничений профиля
    private static final int EXACT = 0;  // точная отметка (выше/ниже коммуникации)
    private static final int FLOOR = 1;  // нижняя граница (не мельче)

    /** Требуемые отметки на пересечениях: [startM, endM, depthM, type]. */
    private static List<double[]> crossingConstraints(NewSegment seg, RefData ref,
                                                      List<Model.ConstraintZone> specialZones,
                                                      ConstraintGrid grid) {
        LineString line = GF.createLineString(seg.coords.toArray(new Coordinate[0]));
        double length = line.getLength();
        double base = targetDepthM(ref);
        double minD = ref.depthRule("min_depth_m", 0.7);
        double flat = ref.depthRule("crossing_flat_m", 4.0);
        List<double[]> out = new ArrayList<>();
        double envH = ref.envelopeHeightM(seg.diameterMm);

        double slope = Math.max(ref.depthRule("max_slope", 0.10), 1e-6);
        org.locationtech.jts.linearref.LengthIndexedLine lil =
                new org.locationtech.jts.linearref.LengthIndexedLine(line);
        for (Model.ConstraintZone z : specialZones) {
            Map<String, Object> rule = ref.restrictionRule(z.restrictionType);
            if (rule == null) {
                continue;
            }
            Object vert = rule.get("vertical_clearance_m");
            Object under = rule.get("under_clearance_m");
            if (vert == null && under == null) {
                continue;
            }
            boolean hits = z.geom.distance(line) <= 1e-9;
            double d;      // ось препятствия вдоль линии
            double a, b;   // окно обязательной отметки
            if (hits) {
                Geometry inter = line.intersection(z.geom);
                List<double[]> pts = ConstraintGrid.intervalPoints(inter);
                if (pts.isEmpty()) {
                    continue;
                }
                double dMin = Double.POSITIVE_INFINITY, dMax = Double.NEGATIVE_INFINITY;
                for (double[] p : pts) {
                    double dd = lil.indexOf(new Coordinate(p[0], p[1]));
                    dMin = Math.min(dMin, dd);
                    dMax = Math.max(dMax, dd);
                }
                d = (dMin + dMax) / 2.0;
                a = Math.max(0.0, d - flat / 2);
                b = Math.min(length, d + flat / 2);
                // место присоединения к существующей сети — не пересечение
                // (разъяснение №10): вертикальные ограничения там не действуют
                if (grid != null) {
                    double extent = RefData.num(rule.getOrDefault("extent_m", 0.0));
                    if (grid.tapExclusionNear(z.objectId, lil, a, b, extent + 1.0)) {
                        continue;
                    }
                }
            } else {
                // Препятствие за пределами участка: его пандус (уклон ≤ 0,10)
                // может заходить в этот участок через граничный узел — тогда
                // на конце участка требуется та же отметка (§5). Параллельные
                // зоны (ближайшая точка — не конец) игнорируются.
                Coordinate[] np = org.locationtech.jts.operation.distance.DistanceOp
                        .nearestPoints(line, z.geom);
                double endIdx = lil.indexOf(np[0]);
                boolean atEnd = endIdx <= 1e-6 || endIdx >= length - 1e-6;
                if (!atEnd) {
                    continue;
                }
                d = endIdx;
                a = b = endIdx; // точечное требование на границе
            }
            double required; // требуемая отметка
            int type;
            if (under != null) {
                // дорога/трамвай: верх габарита не мельче заданной отметки
                required = quantize(RefData.num(under), ref);
                type = FLOOR;
            } else {
                // выше или ниже существующей коммуникации — что дешевле (§4, таблица 2)
                Map<String, Object> util = ref.existingUtility(z.restrictionType);
                if (util == null) {
                    continue;
                }
                double clearance = RefData.num(vert);
                double existTop = RefData.num(util.get("depth_top_m"));
                double existH = util.get("height_m") != null
                        ? RefData.num(util.get("height_m"))
                        : existingEnvelopeHeight(z, ref);
                double above = quantize(existTop - clearance - envH, ref); // верх новой выше низа существующей
                double below = quantize(existTop + existH + clearance, ref); // верх новой ниже низа существующей
                if (above < minD + 1e-9) {
                    required = below; // «выше» физически невозможно — только «ниже»
                } else {
                    required = ref.depthK(above) <= ref.depthK(below) ? above : below;
                }
                type = EXACT;
            }
            if (type == EXACT && Math.abs(required - base) <= 1e-9) {
                continue;
            }
            if (!hits) {
                // пандус дотягивается до этого участка только в пределах длины спуска
                double rampLen = Math.abs(required - base) / slope + flat / 2;
                double gap = z.geom.distance(line);
                if (gap > rampLen + 1e-9) {
                    continue;
                }
            }
            out.add(new double[]{a, b, required, type});
        }
        return out;
    }

    /** Габарит существующей тепловой сети — по таблице 1 (ДУ из параметров зоны). */
    private static double existingEnvelopeHeight(Model.ConstraintZone z, RefData ref) {
        Object dn = z.params != null ? z.params.get("diameter") : null;
        double dnMm = dn instanceof Number ? ((Number) dn).doubleValue() : 500.0;
        return ref.envelopeHeightM(dnMm);
    }

    /**
     * Продольный профиль: [distanceM, depthM] с пандусами ≤ maxSlope.
     * Точные отметки (выше/ниже коммуникаций) соединяются с базовой глубиной
     * пандусами ограниченного уклона; между близкими препятствиями изменённая
     * глубина сохраняется (§5). Нижние границы (дороги/трамваи) удерживаются
     * как «не мельче».
     */
    private static List<double[]> profile(double length, List<double[]> constraints,
                                          Double entryDepthM, RefData ref) {
        double base = targetDepthM(ref);
        double slope = Math.max(ref.depthRule("max_slope", 0.10), 1e-6);
        double step = 2.0;
        int n = Math.max(2, (int) (length / step) + 1);
        double[] xs = new double[n];
        for (int i = 0; i < n; i++) {
            xs[i] = i * length / (n - 1);
        }
        double[] prof = new double[n];
        java.util.Arrays.fill(prof, base);
        boolean[] pin = new boolean[n]; // точные отметки неподвижны (§5)
        if (entryDepthM != null) {
            prof[0] = quantize(entryDepthM, ref);
            pin[0] = true;
        }
        for (double[] c : constraints) {
            if (c[3] != EXACT) {
                continue;
            }
            for (int i = 0; i < n; i++) {
                if (c[0] - 1e-9 <= xs[i] && xs[i] <= c[1] + 1e-9) {
                    if (c[2] > base) {
                        prof[i] = Math.max(prof[i], c[2]);
                    } else {
                        prof[i] = pin[i] ? Math.min(prof[i], c[2]) : c[2];
                    }
                    pin[i] = true;
                }
            }
        }

        // сглаживание уклоном: прямой и обратный проход; закреплённые точки
        // не трогаем — иначе прямой проход «съедает» пандус (восстановление
        // после обратного прохода зацикливается без распространения влево)
        for (int iter = 0; iter < 8; iter++) {
            for (int i = 1; i < n; i++) {
                if (pin[i]) {
                    continue;
                }
                double s = slope * (xs[i] - xs[i - 1]);
                prof[i] = prof[i] > prof[i - 1]
                        ? Math.min(prof[i], prof[i - 1] + s)
                        : Math.max(prof[i], prof[i - 1] - s);
            }
            for (int i = n - 2; i >= 0; i--) {
                if (pin[i]) {
                    continue;
                }
                double s = slope * (xs[i + 1] - xs[i]);
                prof[i] = prof[i] > prof[i + 1]
                        ? Math.min(prof[i], prof[i + 1] + s)
                        : Math.max(prof[i], prof[i + 1] - s);
            }
            // нижние границы (дороги/трамваи) удерживаем после каждого прохода
            for (double[] c : constraints) {
                if (c[3] != FLOOR) {
                    continue;
                }
                for (int i = 0; i < n; i++) {
                    if (c[0] - 1e-9 <= xs[i] && xs[i] <= c[1] + 1e-9) {
                        prof[i] = Math.max(prof[i], c[2]);
                    }
                }
            }
        }
        List<double[]> out = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            out.add(new double[]{xs[i], quantize(prof[i], ref)});
        }
        return out;
    }

    /**
     * Построить профиль §5, разрезать участок на отметке 3,0 м, посчитать Kгл.
     * Возвращает подучастки с depthStart/depthEnd и служебным coords3d.
     */
    public static List<NewSegment> applyDepth(NewSegment seg, RefData ref, Map<String, Integer> seq,
                                              List<TechnicalNode> techNodes,
                                              List<Model.ConstraintZone> specialZones,
                                              Double entryDepthM) {
        return applyDepth(seg, ref, seq, techNodes, specialZones, entryDepthM, null);
    }

    /** Вариант с учётом точек присоединения (разъяснение №10). */
    public static List<NewSegment> applyDepth(NewSegment seg, RefData ref, Map<String, Integer> seq,
                                              List<TechnicalNode> techNodes,
                                              List<Model.ConstraintZone> specialZones,
                                              Double entryDepthM, ConstraintGrid grid) {
        LineString line = GF.createLineString(seg.coords.toArray(new Coordinate[0]));
        double length = line.getLength();
        double base = targetDepthM(ref);
        List<double[]> constraints = crossingConstraints(seg, ref, specialZones, grid);
        List<double[]> prof = profile(length, constraints, entryDepthM, ref);

        // точки деления: переходы «базовая отметка ↔ отклонение» (начало и
        // конец пандуса — §5: пересечение отметки 3,0 м = технический узел),
        // а также пересечение 3,0 м в середине пандуса (спуск «с мелкой на
        // глубокую»), чтобы каждый подучасток был равномерным по глубине.
        TreeSet<Double> cuts = new TreeSet<>();
        cuts.add(0.0);
        double tol = 5e-3;
        for (int i = 1; i < prof.size(); i++) {
            double h0 = prof.get(i - 1)[1], h1 = prof.get(i)[1];
            boolean atBase0 = Math.abs(h0 - base) <= tol;
            boolean atBase1 = Math.abs(h1 - base) <= tol;
            if (atBase0 != atBase1) {
                cuts.add(atBase0 ? prof.get(i - 1)[0] : prof.get(i)[0]);
            } else if (!atBase0 && (h0 - base) * (h1 - base) < 0) {
                double x0 = prof.get(i - 1)[0], x1 = prof.get(i)[0];
                double t = (base - h0) / (h1 - h0);
                cuts.add(x0 + t * (x1 - x0));
            }
        }
        cuts.add(length);
        List<Double> cutList = new ArrayList<>();
        for (double c : cuts) {
            cutList.add(Math.rint(c * 100.0) / 100.0);
        }
        cutList = new ArrayList<>(new TreeSet<>(cutList));

        List<NewSegment> parts = new ArrayList<>();
        String curStartNode = seg.startNodeId;
        for (int k = 0; k < cutList.size() - 1; k++) {
            double a = cutList.get(k), b = cutList.get(k + 1);
            if (b - a < 0.01) {
                continue;
            }
            LineString piece = Hydraulics.subline(line, a, b);
            double dA = depthAt(prof, a);
            double dB = depthAt(prof, b);
            boolean isLast = k == cutList.size() - 2;
            String endNode;
            if (isLast) {
                endNode = seg.endNodeId;
            } else {
                seq.merge("tech", 1, Integer::sum);
                endNode = "node-" + seq.get("tech");
                org.locationtech.jts.linearref.LengthIndexedLine lil =
                        new org.locationtech.jts.linearref.LengthIndexedLine(line);
                Coordinate pc = lil.extractPoint(b);
                techNodes.add(new TechnicalNode(endNode,
                        GF.createPoint(pc), "пересечение отметки 3,0 м (§5)"));
            }
            NewSegment part = new NewSegment(
                    k == 0 && isLast ? seg.objectId : seg.objectId + "." + (k + 1),
                    new ArrayList<>(List.of(piece.getCoordinates())),
                    seg.flowTph, seg.role, seg.method, seg.kSpecial,
                    curStartNode, endNode);
            part.diameterMm = seg.diameterMm;
            part.warnings = new ArrayList<>(seg.warnings);
            part.depthStart = Math.rint(dA * 100.0) / 100.0;
            part.depthEnd = Math.rint(dB * 100.0) / 100.0;
            // §5: на равномерном спуске/подъёме — среднее арифметическое Kгл
            double kAvg = (ref.depthK(dA) + ref.depthK(dB)) / 2.0;
            part.tariffRubM = seg.tariffRubM;
            part.costRub = seg.costRub * (part.lengthM / Math.max(seg.lengthM, 1e-9)) * kAvg;
            part.coords3d = coords3d(piece, prof, a, b);
            parts.add(part);
            curStartNode = endNode;
        }

        if (parts.isEmpty()) { // вырожденный случай — участок как есть
            seg.depthStart = base;
            seg.depthEnd = base;
            List<double[]> c3 = new ArrayList<>();
            for (Coordinate c : seg.coords) {
                c3.add(new double[]{c.x, c.y, -base});
            }
            seg.coords3d = c3;
            parts.add(seg);
        }
        return parts;
    }

    /** Глубина профиля в точке x (линейная интерполяция). */
    private static double depthAt(List<double[]> prof, double x) {
        if (x <= prof.get(0)[0]) {
            return prof.get(0)[1];
        }
        if (x >= prof.get(prof.size() - 1)[0]) {
            return prof.get(prof.size() - 1)[1];
        }
        for (int i = 1; i < prof.size(); i++) {
            if (prof.get(i)[0] >= x) {
                double x0 = prof.get(i - 1)[0], h0 = prof.get(i - 1)[1];
                double x1 = prof.get(i)[0], h1 = prof.get(i)[1];
                double t = (x - x0) / Math.max(x1 - x0, 1e-9);
                return h0 + t * (h1 - h0);
            }
        }
        return prof.get(prof.size() - 1)[1];
    }

    /** Служебные 3D-координаты куска (в GeoJSON не выгружаются, §5). */
    private static List<double[]> coords3d(LineString piece, List<double[]> prof,
                                           double a, double b) {
        List<double[]> out = new ArrayList<>();
        double len = piece.getLength();
        org.locationtech.jts.linearref.LengthIndexedLine lil =
                new org.locationtech.jts.linearref.LengthIndexedLine(piece);
        for (Coordinate c : piece.getCoordinates()) {
            double d = lil.indexOf(c);
            out.add(new double[]{c.x, c.y, -depthAt(prof, a + (len > 1e-9 ? d : 0.0))});
        }
        return out;
    }
}
